_posts/2020-05-20-hive-acid.md

```diff
@@ -0,0 +1,317 @@
+>These traits have long been expected of database systems as part of their
+>transaction functionality.“
+
+Several concepts like transactions, writeIds, deltas, locks, etc are added in
+Hive to achieve ACID semantics which we will cover later in the blog.
+
+# Need for Hive ACID Tables
+In any organisation, there is always a need to update or delete existing entries
+in tables e.g. a user writes or updates the review for the service availed a
+week back or a transaction status is changed after a day, etc.
+With regulations like GDPR/CCPA updates/deletes become even more frequent as the
+users can ask the organisation to delete the data on them and organisations are
+obligated to fulfill these requests.
+
+The standard practice to update data has been to overwrite the partition or
+table with the updated data but this is inefficient and unreliable. It takes a
+lot of resources to overwrite all of the existing data to update a few entries
+but more importantly there are issues around isolation when reads on old data
+are going on and the overwrite starts deleting that data. To solve these issues
+several solutions have been developed,many of them are covered
+[in this blog](https://www.qubole.com/blog/qubole-open-sources-multi-engine-support-for-updates-and-deletes-in-data-lakes/),
+and Hive ACID is one of them.
+
+# Concepts of Hive ACID tables
+
+## Types of Hive ACID tables
+There are two types of Hive ACID tables: Insert-Only and Full CRUD ACID tables.
+Following table compares the two:
+
+| Type of ACID Table     | Hive DML Operations Supported | Input Formats supported | Synthetic columns in file? | Additional Table Properties   |
+|:---------------------:|:--------------------------:|:-------------------------:|:----------------------------:|:-----------------------------:|
+|Insert Only ACID Tables| INSERT                     | All input formats         |    No                        | 'transactional'='true', 'transactional_properties'='insert_only' |
+|Full CRUD ACID Tables  | INSERT, UPDATE, DELETE     | ORC                       |    Yes                       | 'transactional'='true' |
+
+## Hive Transactions
+Hive ACID Tables should be accessed under Hive Transactions only. Note that
+these transactions are different from Presto transactions and are managed by
+Hive. Running DML queries under separate transactions helps in atomicity as each
+transaction is independent and when rolled back will not have any impact on the
+state of the table.
+
+## WriteIds
+DML queries under a transaction write to a unique location under partition/table
+described in detail later in Delta Directories section. This location is derived
+by writeId allocated to the transaction. This provides Isolation of DML queries
+and such queries can run in parallel, whenever they can, without interfering
+with each other.
+
+## Valid WriteIds
+DQL queries under a transaction get a list of valid writeIds that belong to the
```

Comment:
I don't think that DQL is a common term -- I actually had to look it up -- whereas DML (and DDL) is used frequently. It might be easier to understand if we replace it with "read", i.e. say "read queries" everywhere.

> Read queries under a transaction ...

> DML queries take a write-lock ... while read queries take a read-lock ...